_posts/2023-11-24-arcanization.md

```diff
@@ -0,0 +1,72 @@
+
+Interestingly, `wgpu` also had to maintain internal reference counts to resources, to keep track of the dependencies between them (for example a bind group depends on the bindings it refers to). This reference counting was carried out manually, and rather error-prone.
+
+# The solution
+
+"Arcanization", as it names implies, was the process of moving resources behind atomic reference counted pointers (`Arc<T>`). Today the `Hub` still holds resource arrays, however these contain `Arc`s instead of the data directly. This lets us hold the locks for much shorter times - in a lot of cases only while cloning the arc - which can then be read from safely outside of the critical section. In addition, some areas of the code don't need to hold locks once the reference has been extracted.
+
+![A simplified diagram showing resources stored via Arcs](img/arcanization-after.png)
+
+The result is much lower lock contention. If you use `wgpu` from multiple threads, this should significantly improve performance. Our friends in the [bevy engine][bevy] community noted that some initial testing showed that with arcanization, the encoding of  shadow-related commands can run in parallel with the main passes, yielding 45% frame time reduction on a test scene (the famous bistro scene) compared to their single threaded configuration. Without arcanization, lock contention is too high to significantly improve performance.
+
+In addition, `wgpu`'s internals are now simpler. This change lifted some restrictions and opens the door for further performance and ergonomics improvements.
+
+# `wgpu` 0.19
+
+The next release featuring this work will be `0.19.0` which we expect to publish around January 17th. We made sure to merge the changes early in the release cycle to give ourselves as much time as we can to catch potential regressions.
+
+This is an absolutely massive change and while we have and are testing as best we can, we do need help from everyone else. Please try updating your project to the latest wgpu and running it. Please report any issues you find!
+
+# What's next?
+
+## Lifting `RenderPass<'a>` lifetime restrictions
+
+If you have used `wgpu`, there is decent chance that you have had to work around the restrictions imposed by the `'rpass` lifetime in a lot of `RenderPass`'s methods, such as `set_bind_group`, `set_pipeline`, and, `set_vertex_buffer`. The recent changes give us the opportunity to store `Arc`s where `&'a` references were previously needed which should let us remove these lifetime restrictions. A [pull request](github.com/gfx-rs/wgpu/pull/4768) is already open to address this.
```

Comment:
We shouldn't mention this PR, it doesn't actually fix the problem, just removes the lifetimes.